src/chainstate/burn/db/burndb.rs

```diff
@@ -604,16 +681,268 @@ impl BurnDB {
+    }
+
+    /// Get an ancestor block snapshot if the given ancestor hash is indeed an ancestor.
+    pub fn get_ancestor_snapshot_by_hash<'a>(tx: &mut BurnDBTx<'a>, ancestor_hash: &BurnchainHeaderHash, tip_hash: &BurnchainHeaderHash) -> Result<Option<BlockSnapshot>, db_error> {
+        let _ = BurnDB::get_block_snapshot(tx, tip_hash)?.ok_or(db_error::NotFoundError)?;
+        let an_sn = match BurnDB::get_block_snapshot(tx, ancestor_hash)? {
+            Some(sn) => sn,
+            None => {
+                return Ok(None);
+            }
+        };
+
+        match burndb_get_ancestor_block_hash(tx, an_sn.block_height, tip_hash)? {
+            Some(bhh) => {
+                if bhh != an_sn.burn_header_hash {
+                    return Ok(None);
+                }
+                else {
+                    return Ok(Some(an_sn));
+                }
+            },
+            None => {
+                return Ok(None);
+            }
+        }
+    }
+
+    /// Get a snapshot with an arrived block (i.e. a block that was marked as processed)
+    fn get_snapshot_by_arrival_index(conn: &Connection, arrival_index: u64) -> Result<Option<BlockSnapshot>, db_error> {
+        let mut rows = query_rows(conn, "SELECT * FROM snapshots WHERE arrival_index = ?1 AND stacks_block_accepted > 0", &[&u64_to_sql(arrival_index)?])?;
+        let len = rows.len();
+        match len {
+            0 => Ok(None),
+            1 => Ok(rows.pop()),
+            _ => {
+                panic!("BUG: multiple snapshots have the same non-zero arrival index");
+            }
+        }
+    }
+
+    /// Find all stacks blocks that were processed since parent_tip had been processed, and generate MARF
+    /// key/value pairs for the subset that arrived on ancestor blocks of the parent.  Update the
+    /// given parent chain tip to have the correct memoized canonical chain tip present in the fork
+    /// it represents.
+    fn process_new_block_arrivals<'a>(tx: &mut BurnDBTx<'a>, parent_tip: &mut BlockSnapshot) -> Result<(Vec<String>, Vec<String>), db_error> {
+        let mut keys = vec![];
+        let mut values = vec![];
+
+        let max_arrival_index_key = BurnDB::make_stacks_block_max_arrival_index_key();
+        let old_max_arrival_index_str = BurnDB::index_value_get(tx, &parent_tip.burn_header_hash, &max_arrival_index_key)?.unwrap_or("0".to_string());
+        let old_max_arrival_index = old_max_arrival_index_str.parse::<u64>().expect("BUG: max arrival index is not a u64");
+        let max_arrival_index = BurnDB::get_arrival_index(tx)?;
+
+        let mut new_block_arrivals = vec![];
+
+        // find all Stacks block hashes who arrived since this parent_tip was built.
+        for ari in old_max_arrival_index..(max_arrival_index+1) {
+            let arrival_sn = match BurnDB::get_snapshot_by_arrival_index(tx, ari)? {
+                Some(sn) => sn,
+                None => {
```

Comment:
What would cause this to return `None`?

Reply:
This can happen only when the database is being initialized -- i.e. before the first `BlockSnapshot` gets inserted.